File: ros_ws/src/navigation_collisions/src/engine_shapes.py
import math
import time

from collisions_shapes_attrib import Point

class MapObstacle(object):
    def __init__(self, position, velocity = None):
        self.Position = position
        self.Velocity = velocity
        self.SpawnTime = time.time()


class CircleObstacle(MapObstacle):
    def __init__(self, position, radius, velocity = None):
        super(CircleObstacle, self).__init__(position, velocity)
        self.Radius = radius

    def __repr__(self):
        return "circle"


class RectObstacle(MapObstacle):
    def __init__(self, position, width, height, velocity = None):
        super(RectObstacle, self).__init__(position, velocity)
        self.Width = width
        self.Height = height

    def corners(self): # Calculs persos, a verifier DEPRECATED
        corners = []
        l = math.sqrt((self.Width / 2.0) ** 2 + (self.Height / 2.0) ** 2)
        corner_phi = math.atan2(self.Height, self.Width)
        for angle_phi in [0, math.pi]:
            for i in range(2):
                phi = angle_phi  + ((-1) ** (i + 1)) * corner_phi
                corners.append(Point(self.Position.X + l * math.cos(phi + self.Position.A), self.Position.Y + l * math.sin(phi + self.Position.A)))
        return corners

    def segments(self):
        c = self.corners()
        return [(c[i], c[(i + 1) % 4]) for i in range(len(c))]

    def __repr__(self):
        return "rect"

File: ros_ws/src/navigation_coppisions/src/collisions_engine/engine_shapes.cpp
#include "engine_shapes.h"
#include <math.h>

MapObstacle::MapObstacle() //TODO good c++ ..?
{
    MapObstacle::position = Position(0, 0, 0);
    MapObstacle::velocity = Velocity(0, 0);
    float spawn_time = 0;
}
MapObstacle::MapObstacle(Point point, Velocity velocity = nullptr)
{
    MapObstacle::position = Position(point.x, point.y, 0);
    MapObstacle::velocity = velocity;
    float spawn_time = 0;
}
MapObstacle::MapObstacle(Position position, Velocity velocity = nullptr)
{
    MapObstacle::position = position;
    MapObstacle::velocity = velocity;
    float spawn_time = 0;
}


SegmentObstacle::SegmentObstacle(Point first_point, Point last_point, Velocity velocity = nullptr)
    :MapObstacle(first_point, velocity)
{
    ObstacleType type = ObstacleType::SEGMENT;
    first = first_point;
    last  = last_point;

    length = sqrt(pow(last.x - first.x, 2) + pow(last.y - last.y, 2));
    position = Position((first.x + last.x) / 2, (first.y + last.y) / 2, 
                        atan2(last.y - first.y, last.x - first.x)); // center position
}


CircleObstacle::CircleObstacle(Position position, float radius, Velocity velocity = nullptr)
    :MapObstacle(position, velocity)
{
    ObstacleType type = ObstacleType::CIRCLE;
    CircleObstacle::radius = radius;
}


RectObstacle::RectObstacle(Position position, float width, float height, Velocity velocity = nullptr)
    :MapObstacle(position, velocity)
{
    ObstacleType type = ObstacleType::RECT;
    RectObstacle::width = width;
    RectObstacle::height = height;
}

std::vector<Point> RectObstacle::get_corners()
{
    std::vector<Point> corners;
    float l = sqrt(pow(width / 2.0, 2) + pow(height / 2.0, 2));
    float corner_phi = atan2(height, width);

    float angle_phi = 0;
    for(int j = 0; j < 2; j++) {
        for(int i = 0; i < 2; i++)
        {
            float phi = angle_phi + pow(i + 1, -1) * corner_phi;
            corners.push_back(Point(position.x + l * cos(phi + position.a), 
                                    position.y + l * sin(phi + position.a)));
        }
        angle_phi += M_PI;
    }
    return corners;
}

std::vector<SegmentObstacle> RectObstacle::get_segments()
{
    std::vector<Point> corners = get_corners();

    std::vector<SegmentObstacle> segments;
    for(int i = 0; i < 4; i++)
        segments.push_back(SegmentObstacle(corners[i], corners[(i + 1) % 4]));
    
    return segments;
}